GPTutils.py

```diff
@@ -0,0 +1,262 @@
+from typing import List, Dict, Tuple, Union
+import asyncio
+
+from openai import OpenAI
+import openai
+import langchain
+from langchain.embeddings import OpenAIEmbeddings
+from langchain.callbacks.streaming_aiter_final_only import (
+    AsyncFinalIteratorCallbackHandler,
+)
+from langchain.text_splitter import (
+    RecursiveCharacterTextSplitter,
+    CharacterTextSplitter,
+)
+import numpy as np
+from numpy.linalg import norm
+from GPTsettings import GPTsettings
+
+mapper = {
+    "py": "python",
+    "js": "js",
+    "jsx": "js",
+    "ts": "ts",
+    "java": "java",
+    "html": "html",
+    "md": "markdown",
+}
+
+
+class GPTWrapper:
+    def __init__(self, model: str = GPTsettings.MODEL) -> None:
+        """
+        Initializes a new instance of the GPTWrapper class.
+
+        Args:
+            model (str, optional): The name of the GPT model to use. Defaults to "text-davinci-002".
+        """
+        self.client = OpenAI()
+        self.model: str = model
+        self.conversation: list[
+            dict[str, str]
+        ] = []  # This will hold our conversation messages
+
+    def add_message(self, role: str, content: str) -> None:
+        """
+        Adds a message to the conversation history.
+
+        Args:
+            role (str): Indicates who is acting on the conversation [system, user, assistant].
+            content (str): The actual message.
+        """
+        self.conversation.append({"role": role, "content": content})
+
+    def edit_message(self, index: int, role: str, content: str) -> None:
+        """
+        Edits a message in the conversation by index.
+
+        Args:
+            index (int): The index of the message to edit.
+            role (str): Indicates who is acting on the conversation [system, user, assistant].
+            content (str): The new content of the message.
+
+        Raises:
+            IndexError: If the index is out of range.
+        """
+        if index < len(self.conversation):
+            self.conversation[index] = {"role": role, "content": content}
+        else:
+            raise IndexError("Message index out of range")
+
+    def get_response(self, max_tokens: int = 1000) -> str:
```

Comment:
Make sure to handle potential exceptions when calling external APIs in `get_response`.